File: src/engine/web/input/gamepad.ts
export type GamepadMapping =
  (typeof GamepadMapping)[keyof typeof GamepadMapping]
const GamepadMapping = {Unknown: 0, Standard: 1} as const

export type GamepadPoll = {
  /** gamepad slot index. */
  index: number
  /** true if the gamepad is still connected. */
  connected: boolean
  /** button layout mapping. */
  mapping: GamepadMapping
  /** bitmask of pressed buttons. */
  buttons: number
  /** axes in [-1, 1]; standard mapping: [leftX, leftY, rightX, rightY]. */
  axes: readonly [number, number, number, number]
}

export class Gamepad {
  /** ;readonly. */
  polls: {[i: number]: GamepadPoll} = {}

  reset(): void {
    this.polls = {}
  }

  update(): void {
    this.polls = {}
    for (const pad of navigator.getGamepads()) {
      if (!pad) continue
      let buttons = 0
      for (let i = 0; i < pad.buttons.length; i++)
        if (pad.buttons[i]!.pressed) buttons |= 1 << i
      this.polls[pad.index] = {
        index: pad.index,
        connected: pad.connected,
        mapping:
          GamepadMapping[pad.mapping as keyof typeof GamepadMapping] ??
          GamepadMapping.Unknown,
        buttons,
        axes: [
          pad.axes[0] ?? 0,
          pad.axes[1] ?? 0,
          pad.axes[2] ?? 0,
          pad.axes[3] ?? 0
        ]
      }
    }
  }
}

File: src/void/vinput/gamepad.go
package vinput

const maxAxes = 4 // leftX, leftY, rightX, rightY.

type GamepadMapping uint8

const (
	GamepadMappingUnknown GamepadMapping = iota
	GamepadMappingStandard
)

type GamepadButton uint16

const (
	GamepadButtonA GamepadButton = 1 << iota // cross.
	GamepadButtonB                           // circle.
	GamepadButtonX                           // square.
	GamepadButtonY                           // triangle.
	GamepadButtonL1
	GamepadButtonR1
	GamepadButtonL2
	GamepadButtonR2
	GamepadButtonSelect
	GamepadButtonStart
	GamepadButtonL3
	GamepadButtonR3
	GamepadButtonUp
	GamepadButtonDown
	GamepadButtonLeft
	GamepadButtonRight

	gamepadButtonBits = iota
)

// gamepad state. index matches the browser gamepad index slot.
type GamepadPoll struct {
	// gamepad slot index.
	Index uint8
	// true if the gamepad is still connected.
	Connected bool
	// button layout mapping.
	Mapping GamepadMapping
	// bitmask of pressed buttons.
	Buttons GamepadButton
	// axes in [-1, 1]; standard mapping: [leftX, leftY, rightX, rightY].
	Axes [4]float32
}

type Gamepad struct {
	GamepadPoll
}
